staging/src/k8s.io/client-go/rest/request.go

```diff
@@ -455,24 +455,67 @@ func (r *Request) URL() *url.URL {
 
 // finalURLTemplate is similar to URL(), but will make all specific parameter values equal
 // - instead of name or namespace, "{name}" and "{namespace}" will be used, and all query
-// parameters will be reset. This creates a copy of the request so as not to change the
-// underlying object.  This means some useful request info (like the types of field
-// selectors in use) will be lost.
-// TODO: preserve field selector keys
+// parameters will be reset. This creates a copy of the url so as not to change the
+// underlying object.
 func (r Request) finalURLTemplate() url.URL {
-	if len(r.resourceName) != 0 {
-		r.resourceName = "{name}"
-	}
-	if r.namespaceSet && len(r.namespace) != 0 {
-		r.namespace = "{namespace}"
-	}
 	newParams := url.Values{}
 	v := []string{"{value}"}
 	for k := range r.params {
 		newParams[k] = v
 	}
 	r.params = newParams
 	url := r.URL()
+	segments := strings.Split(r.URL().Path, "/")
+	groupIndex := 0
+	index := 0
+	if r.URL() != nil && r.baseURL != nil && strings.Contains(r.URL().Path, r.baseURL.Path) {
+		groupIndex += len(strings.Split(r.baseURL.Path, "/"))
+	}
+	if groupIndex >= len(segments) {
+		return *url
+	}
+
+	const CoreGroupPrefix = "api"
+	const NamedGroupPrefix = "apis"
+	isCoreGroup := segments[groupIndex] == CoreGroupPrefix
+	isNamedGroup := segments[groupIndex] == NamedGroupPrefix
+	if isCoreGroup {
+		// checking the case of core group with /api/v1/... format
+		index = groupIndex + 2
+	} else if isNamedGroup {
+		// checking the case of named group with /apis/apps/v1/... format
+		index = groupIndex + 3
+	} else {
+		// this should not happen that the only two possibilities are /api... and /apis..., just want to put an
+		// outlet here in case more API groups are added in future if ever possible:
+		// https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-groups
+		r.err = fmt.Errorf("Request has a invald Group name %s", segments[groupIndex])
+		return *url
+	}
+	//switch segLength := len(segments) - index; segLength {
+	switch {
+	// case len(segments) - index == 2:
+	// resource (with no name) do nothing
+	case len(segments)-index == 2:
+		// /$RESOURCE/$NAME: replace $NAME with {name}
+		segments[index+1] = "{name}"
+	case len(segments)-index == 3:
+		if segments[index+2] == "finalize" || segments[index+2] == "status" {
+			// /$RESOURCE/$NAME/$SUBRESOURCE: replace $NAME with {name}
+			segments[index+1] = "{name}"
+		} else {
+			// /namespace/$NAMESPACE/$RESOURCE: replace $NAMESPACE with {namespace}
+			segments[index+1] = "{namespace}"
+		}
+	case len(segments)-index >= 4:
+		segments[index+1] = "{namespace}"
+		// /namespace/$NAMESPACE/$RESOURCE/$NAME: replace $NAMESPACE with {namespace},  $NAME with {name}
+		if segments[index+3] != "finalize" && segments[index+3] != "status" {
```

Comment:
Should be `index+2`? It looks like you are distinguishing

 - /namespaces/{namespace}/pods/{name}
 - /namespaces/{name}/finalize/subpath

Although I don't think subresources of the namespace Kind (finalize and status) have subpaths.

Reply:
This is to distinguish between:
/namespace/$NAMESPACE/$RESOURCE/$NAME
/namespace/$NAMESPACE/$RESOURCE/$SUBRESOURCE

so if segments[index+3] is not "finalize" or "status", then it's $NAME, replace it with {name}:
`segments[index+3] = "{name}"`

Right?

Reply:
That's brittle since we could add new subresources for namespaces in the future, but the consequence of getting it wrong is somewhat minor, so I'm inclined to say just go ahead and do that. (be sure to add a test case)

Reply:
I am sorry, what are you referring to by "do that"?

Reply:
Reserving "finalize" and "status" as keywords is brittle to future API changes for namespaces, but currently it seems to be the only way for a rest client (with all the information it has) to distinguish between the following two valid requests:

 - `GET /api/v1/namespaces/default/pods`: replace "default" with {namespace}
 - `GET /api/v1/namespaces/default/status`: replace "default" with {name}

I prefer forcing how clients are used rather than just parsing the requests (as a followup)

> This is to distinguish between:
/namespace/$NAMESPACE/$RESOURCE/$NAME
/namespace/$NAMESPACE/$RESOURCE/$SUBRESOURCE

$SUBRESOURCE isn't constrained to "finalize" and "status" for Kinds other than namespaces

> /namespaces/{name}/finalize/subpath
> /namespaces/$NAMESPACE/$RESOURCE/$SUBRESOURCE

For the two cases above, they don't exist in Kubernetes yet, I suggest we don't handle/test them.

Reply:
Thanks for the explanation! I will definitely take all these into consideration when I implement a better approach after code freeze for the next release.